chia/util/streamable.py

```diff
@@ -152,80 +155,121 @@ def streamable(cls: Any):
-            else:
-                raise ValueError("Optional must be 0 or 1")
+            parse_inner_type_f = cls.function_to_parse_one_item(inner_type)
+            return lambda f: parse_optional(f, parse_inner_type_f)
         if hasattr(f_type, "parse"):
-            return f_type.parse(f)
+            return f_type.parse
         if f_type == bytes:
-            list_size_bytes = f.read(4)
-            assert list_size_bytes is not None and len(list_size_bytes) == 4  # Checks for EOF
-            list_size: uint32 = uint32(int.from_bytes(list_size_bytes, "big"))
-            bytes_read = f.read(list_size)
-            assert bytes_read is not None and len(bytes_read) == list_size
-            return bytes_read
+            return parse_bytes
         if is_type_List(f_type):
             inner_type = get_args(f_type)[0]
-            full_list: List[inner_type] = []  # type: ignore
-            # wjb assert inner_type != get_args(List)[0]  # type: ignore
-            list_size_bytes = f.read(4)
-            assert list_size_bytes is not None and len(list_size_bytes) == 4  # Checks for EOF
-            list_size = uint32(int.from_bytes(list_size_bytes, "big"))
-            for list_index in range(list_size):
-                full_list.append(cls.parse_one_item(inner_type, f))  # type: ignore
-            return full_list
+            parse_inner_type_f = cls.function_to_parse_one_item(inner_type)
+            return lambda f: parse_list(f, parse_inner_type_f)
         if is_type_Tuple(f_type):
             inner_types = get_args(f_type)
-            full_list = []
-            for inner_type in inner_types:
-                full_list.append(cls.parse_one_item(inner_type, f))  # type: ignore
-            return tuple(full_list)
+            list_parse_inner_type_f = [cls.function_to_parse_one_item(_) for _ in inner_types]
+            return lambda f: parse_tuple(f, list_parse_inner_type_f)
         if hasattr(f_type, "from_bytes") and f_type.__name__ in size_hints:
             bytes_to_read = size_hints[f_type.__name__]
-            bytes_read = f.read(bytes_to_read)
-            assert bytes_read is not None and len(bytes_read) == bytes_to_read
-            return f_type.from_bytes(bytes_read)
+            return lambda f: parse_size_hints(f, f_type, bytes_to_read)
         if f_type is str:
-            str_size_bytes = f.read(4)
-            assert str_size_bytes is not None and len(str_size_bytes) == 4  # Checks for EOF
-            str_size: uint32 = uint32(int.from_bytes(str_size_bytes, "big"))
-            str_read_bytes = f.read(str_size)
-            assert str_read_bytes is not None and len(str_read_bytes) == str_size  # Checks for EOF
-            return bytes.decode(str_read_bytes, "utf-8")
-        raise RuntimeError(f"Type {f_type} does not have parse")
+            return parse_str
+        raise NotImplementedError(f"Type {f_type} does not have parse")
```

Comment:
this message doesn't look right to me. Shouldn't it say "Type {f_type} does not have from_bytes()"?

Reply:
I see, `parse()` is the main way of supporting the streamable protocol

Reply:
Currently `from_bytes()` is used only for sized integer types. All other types that have custom serialization need to provide their own `.parse()` function.